helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/AbstractPartitionMovementConstraint.java

```diff
@@ -0,0 +1,86 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.util.Collections;
+import java.util.Map;
+
+import org.apache.helix.controller.rebalancer.waged.model.AssignableNode;
+import org.apache.helix.controller.rebalancer.waged.model.AssignableReplica;
+import org.apache.helix.controller.rebalancer.waged.model.ClusterContext;
+import org.apache.helix.model.Partition;
+import org.apache.helix.model.ResourceAssignment;
+
+/**
+ * Evaluate the proposed assignment according to the potential partition movements cost.
+ * The cost is evaluated based on the difference between the old assignment and the new assignment.
+ * Any change from the old assignment will increase the partition movements cost, so that the
+ * evaluated score will become lower.
+ */
+abstract class AbstractPartitionMovementConstraint extends SoftConstraint {
+  protected static final double MAX_SCORE = 1f;
+  protected static final double MIN_SCORE = 0f;
+
+  AbstractPartitionMovementConstraint() {
+    super(MAX_SCORE, MIN_SCORE);
+  }
+
+  /**
+   * @return MAX_SCORE if the proposed assignment completely matches the previous assignment.
+   *         StateTransitionCostFactor if the proposed assignment's allocation matches the
+   *         previous assignment but state does not match.
+   *         MIN_SCORE if the proposed assignment completely doesn't match the previous one.
+   */
+  @Override
+  protected abstract double getAssignmentScore(AssignableNode node, AssignableReplica replica,
+      ClusterContext clusterContext);
+
+  /**
+   * @return The scale factor to adjust score when the proposed allocation partially matches the
+   * assignment plan but will require a state transition (with partition movement).
+   */
+  protected abstract double getStateTransitionCostFactor();
+
+  protected Map<String, String> getStateMap(AssignableReplica replica,
+      Map<String, ResourceAssignment> assignment) {
+    String resourceName = replica.getResourceName();
+    String partitionName = replica.getPartitionName();
+    if (assignment == null || !assignment.containsKey(resourceName)) {
+      return Collections.emptyMap();
+    }
+    return assignment.get(resourceName).getReplicaMap(new Partition(partitionName));
+  }
+
+  protected double calculateAssignmentScore(String nodeName, String state,
+      Map<String, String> instanceToStateMap) {
+    if (instanceToStateMap.containsKey(nodeName)) {
+      return state.equals(instanceToStateMap.get(nodeName)) ?
+          MAX_SCORE : // if state matches, no state transition required for the proposed assignment
+          getStateTransitionCostFactor(); // if state does not match, then the proposed assignment
```

Comment:
Should be (MAX_SCORE - MIN_SCORE) * getStateTransitionCostFactor() + MIN_SCORE?

Reply:
I'm reusing the old name, but it's really just the score (0.5), not a scaling factor. I'm renaming the method to clearly indicate that it's returning a score, not a scaling factor.

Reply:
It is OK, but I don't like that design. First of all, I think we all agree that the logic would be the same.
My concern is that factor design is bug free. By returning a number between 0-1, the eventual score would be sure to be between MIN_SCORE and MAX_SCORE. However, if you directly return the getStateTransitionCost(), then we rely on the coder's discipline.

Actually, even you change the method to be getStateTransitionCost(), I think it shall still be implemented with the above formula,
(MAX_SCORE - MIN_SCORE) * getStateTransitionCostFactor() + MIN_SCORE

Reply:
That's a valid point. Ok.